resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosProtoUtils.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler.cluster.mesos
+
+import org.apache.mesos.Protos
+
+import org.apache.spark.SparkException
+import org.apache.spark.internal.Logging
+
+object MesosProtoUtils extends Logging {
+
+  /** Parses a label string of the format specified in spark.mesos.task.labels. */
+  def mesosLabels(labelsStr: String): Protos.Labels.Builder = {
+    var key: Option[String] = None
+    var value: Option[String] = None
+    var currStr = ""
+    var i = 0
+    val labels = Protos.Labels.newBuilder()
+
+    // 0 -> parsing key
+    // 1 -> parsing value
+    var state = 0
+
+    def addLabel() = {
+      value = Some(currStr)
+      if (key.isEmpty) {
+        throw new SparkException(s"Error while parsing label string: ${labelsStr}.  " +
+          s"Empty label key.")
+      } else {
+        val label = Protos.Label.newBuilder().setKey(key.get).setValue(value.get)
+        labels.addLabels(label)
+
+        key = None
+        value = None
+        currStr = ""
+        state = 0
+      }
+    }
+
+    while(i < labelsStr.length) {
```

Comment:
Ideally, this would be more functional.  I tried to model it with map/fold, but I'm not smart enough.  If someone cares, I can try to rewrite it to be recursive, at least.